algebra/Atiyah.lean

```diff
@@ -0,0 +1,1695 @@
+lemma sub_mem : x ∈ S → y ∈ S → x - y ∈ S :=
+λ hx hy, have h : x + -y ∈ S, from add_mem hx $ neg_mem hy, by simpa using h
+
+variables (S x y z)
+
+@[simp, reducible] def r : α → α → Prop :=
+λ x y, x - y ∈ S
+
+lemma refl : r S x x :=
+by simp [sub_self]
+
+lemma symm : r S x y → r S y x :=
+λ hxy, have h : -(x - y) ∈ S, from neg_mem hxy, by simpa using h
+
+lemma trans : r S x y → r S y z → r S x z :=
+λ hxy hyz, have h : (x - y) + (y - z) ∈ S, from add_mem hxy hyz, by simpa using h
+
+instance : setoid α :=
+⟨r S, refl S, symm S, trans S⟩
+
+variables (α)
+
+@[simp, reducible] def coset := quotient (is_ideal.setoid S)
+
+variables {α}
+
+instance : comm_ring (quotient (is_ideal.setoid S)) :=
+by refine
+{ add            := quotient.lift₂ (λ m n, ⟦m + n⟧) (λ m₁ m₂ n₁ n₂ h₁ h₂, quot.sound $ calc
+    (m₁ + m₂) - (n₁ + n₂) = (m₁ - n₁) + (m₂ - n₂) : by norm_num
+                      ... ∈ S                     : add_mem h₁ h₂),
+  add_assoc      := λ m n k, quotient.induction_on₃ m n k _,
+  zero           := ⟦0⟧,
+  zero_add       := quotient.ind _,
+  add_zero       := quotient.ind _,
+  neg            := quotient.lift (λ m, ⟦-m⟧) (λ m n h, quot.sound $ calc
+    (-m) - (-n) = (m - n)*-1 : by norm_num
+            ... ∈ S          : mul_mem h),
+  add_left_neg   := quotient.ind _,
+  add_comm       := quotient.ind₂ _,
+  mul            := quotient.lift₂ (λ m n, ⟦m * n⟧) (λ m₁ m₂ n₁ n₂ h₁ h₂, quot.sound $ calc
+    (m₁ * m₂) - (n₁ * n₂) = (m₂ - n₂) * m₁ + (m₁ - n₁) * n₂ : by ring
+                      ... ∈ S : add_mem (mul_mem h₂) (mul_mem h₁)),
+  mul_assoc      := λ m n k, quotient.induction_on₃ m n k _,
+  one            := ⟦1⟧,
+  one_mul        := quotient.ind _,
+  mul_one        := quotient.ind _,
+  left_distrib   := λ m n k, quotient.induction_on₃ m n k _,
+  right_distrib  := λ m n k, quotient.induction_on₃ m n k _,
+  mul_comm       := quotient.ind₂ _ };
+{ intros, apply quotient.sound, ring }
+
+@[reducible] def to_coset (x : α) : coset α S := quotient.mk x
+
+instance to_quotient : is_hom (to_coset S) :=
+by refine {..}; intros; refl
+
+infix / := coset
+
+variable {S}
+
+@[simp] lemma sub_mem_iff_r : x ≈ y ↔ x - y ∈ S := iff.refl _
+@[simp] lemma sub_mem_iff_coset_eq : ⟦x⟧ = ⟦y⟧ ↔ x - y ∈ S :=
+by rw quotient.eq; exact sub_mem_iff_r x y
+
+variable S
+
+@[simp] lemma add_coset : ⟦x⟧ + ⟦y⟧ = ⟦x + y⟧ := rfl
+@[simp] lemma sub_coset : ⟦x⟧ - ⟦y⟧ = ⟦x - y⟧ := rfl
+@[simp] lemma mul_coset : ⟦x⟧ * ⟦y⟧ = ⟦x * y⟧ := rfl
+@[simp] lemma neg_coset : -⟦x⟧ = ⟦-x⟧ := rfl
+@[simp] lemma one_coset : (1:α/S) = ⟦(1:α)⟧ := rfl
+
+lemma coset_zero_of_mem : x ∈ S → ⟦x⟧ = 0 :=
+λ hx, quotient.sound $ by simp [hx]
+
+lemma mem_of_coset_zero : ⟦x⟧ = 0 → x ∈ S :=
+λ hx, by simpa using quotient.exact hx
+
+lemma mem_iff_coset_zero : x ∈ S ↔ ⟦x⟧ = 0 :=
+⟨coset_zero_of_mem S x, mem_of_coset_zero S x⟩
+
+lemma coset_zero_iff_mem : ⟦x⟧ = 0 ↔ x ∈ S :=
+⟨mem_of_coset_zero S x, coset_zero_of_mem S x⟩
+
```

Comment:
Not sure Mario would like all these...

Reply:
I also would prefer to avoid them. I would suggest to only introduce `coset_zero_iff_mem` and think about tagging it with `@[simp]`. The other versions are just confusing.